### PR TITLE
Hotfix/using html swapper

### PR DESCRIPTION
**Pull request by hythof**
## Purpose
Use translate API on html-swapper server such as `WOVN.php`.

There are features.
- When render default lang or fail communicate to API server (include timeout), Insert snippet and hreflang tags
- Delete CSRF, calculate html hash for Fastly CDN and then restore deleted HTML

## Background
There is latest feature on html-swapper and wovnjava is not.
Because wovnjava is not sufficiently maintained.

### CODE REVIEW

File: src/main/java/com/github/wovnio/wovnjava/Api.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                con = (HttpURLConnection) url.openConnection();
+                con.setConnectTimeout(settings.connectTimeout);
+                con.setReadTimeout(settings.readTimeout);
+            } catch (Exception e) {
+                Logger.log.error("Api url", e);
+                throw new ApiException("unknown");
+            }
+            return translate(lang, html, con);
+        } finally {
+            con.disconnect();
```
Comment (mrasu) [inline src/main/java/com/github/wovnio/wovnjava/Api.java]: `con` can be null if error happens before successing `(HttpURLConnection) url.openConnection()`

File: src/main/java/com/github/wovnio/wovnjava/Api.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    String translate(String lang, String html) throws ApiException {
+        HttpURLConnection con = null;
+        try {
+            try {
+                URL url = getApiUrl(lang, html);
+                con = (HttpURLConnection) url.openConnection();
+                con.setConnectTimeout(settings.connectTimeout);
+                con.setReadTimeout(settings.readTimeout);
+            } catch (Exception e) {
```
Comment (mrasu) [inline src/main/java/com/github/wovnio/wovnjava/Api.java]: please catch specific exceptions you expect.

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java [hunk 3]
```diff
@@ -1,33 +1,112 @@
...
+        } else {
+            tryTranslate(headers, (HttpServletRequest)request, response, chain);
+        }
     }
 
     public void destroy() {
     }
+
+    boolean isHtml(String path) {
+        // TODO: implement
```
Comment (mrasu) [inline src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: todo remains?

File: src/main/java/com/github/wovnio/wovnjava/] [hunk 4]
```diff
@@ -0,0 +1,73 @@
+package com.github.wovnio.wovnjava;
```
Comment (mrasu) [inline src/main/java/com/github/wovnio/wovnjava/]]: Is `src/main/java/com/github/wovnio/wovnjava/]` mistake?

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java (hunk 3 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: The code copy from older one.

File: src/main/java/com/github/wovnio/wovnjava/] (hunk 4 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/]]: Just mistake.
I deleted it.
Thanks.

File: src/main/java/com/github/wovnio/wovnjava/Api.java (hunk 1 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/Api.java]: That's right.
I added null check.

File: src/main/java/com/github/wovnio/wovnjava/Api.java (hunk 2 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/Api.java]: I added catch each exceptions.

File: src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Comment;
+import org.jsoup.nodes.Element;
+import org.jsoup.parser.Tag;
+import org.jsoup.select.Elements;
+
+class HtmlConverter {
+    private final ArrayList<String> markers = new ArrayList<String>();
+    private final String prefix = "wovn-marker-";
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java]: could you please change prefix to CAPITAL and have a better name? like WOVN_MARKER_PREFIX or something else

File: src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        js.attr("data-wovnio", key);
+        js.attr("data-wovnio-type", type);
+        js.attr("async", "async");
+        doc.head().appendChild(js);
+    }
+
+    private void appendHrefLang(Headers headers) {
+        for (String lang : settings.supportedLangs) {
+            Element link = new Element(Tag.valueOf("link"), "");
+            link.attr("ref", "altername");
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java]: `alternate`, not `altername`

File: src/main/java/com/github/wovnio/wovnjava/Interceptor.java [hunk 7]
```diff
@@ -1,608 +1,43 @@
...
-import org.w3c.dom.Element;
-import org.w3c.dom.Node;
-import org.w3c.dom.NodeList;
-import org.w3c.dom.NamedNodeMap;
-
-import nu.validator.htmlparser.dom.*;
-
 class Interceptor {
-    private Store store;
+    private final Settings settings;
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/Interceptor.java]: indent

File: src/main/java/com/github/wovnio/wovnjava/Interceptor.java [hunk 8]
```diff
@@ -1,608 +1,43 @@
...
+	private String apiTranslate(String lang, String body) {
         try {
-            NodeList heads = doc.getElementsByTagName("head");
-            if (heads != null && heads.getLength() > 0) {
-                parentNode = heads.item(0);
-            }
-        } catch (NullPointerException e) {
-            Logger.log.error("NullPointerException while searching <head> tag", e);
+            HtmlConverter converter = new HtmlConverter(settings, body);
+		    String convertedBody = converter.strip();
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/Interceptor.java]: Indent. I think you have mixed spaces and tab here

File: src/main/java/com/github/wovnio/wovnjava/Settings.java [hunk 9]
```diff
@@ -238,4 +225,24 @@ boolean isValid() {
 
         return valid;
     }
+
+    String hash() throws NoSuchAlgorithmException {
+        MessageDigest md = MessageDigest.getInstance("MD5");
+        md.update(projectToken.getBytes());
+        md.update(urlPattern.getBytes());
+        md.update(urlPatternReg.getBytes());
+		for (String q : query) {
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/Settings.java]: indent

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java [hunk 10]
```diff
@@ -1,33 +1,112 @@
...
+import javax.servlet.ServletOutputStream;
 import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
 public class WovnServletFilter implements Filter {
-    protected Interceptor interceptor;
+    //protected Interceptor interceptor;
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: plz remove

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java [hunk 11]
```diff
@@ -1,33 +1,112 @@
...
+    private Settings settings;
 
-    public static final String VERSION = "0.2.1";
+    public static final String VERSION = Settings.VERSION;  // for backword compatibility
```
Comment (suplo) [inline src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: this is not used

File: src/test/java/com/github/wovnio/wovnjava/ApiTest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        mock.setRequestMethod("POST");
+        EasyMock.expect(mock.getResponseCode()).andReturn(code);
+        EasyMock.expect(mock.getContentEncoding()).andReturn(encoding);
+        EasyMock.expect(mock.getOutputStream()).andReturn(new ByteArrayOutputStream());
+        EasyMock.expect(mock.getInputStream()).andReturn(responseStream);
+        EasyMock.replay(mock);
+        return mock;
+    }
+
+    private HttpServletRequest mockRequestPath(String path) {
```
Comment (suplo) [inline src/test/java/com/github/wovnio/wovnjava/ApiTest.java]: duplicate code with InterceptorTest.java

File: src/test/java/com/github/wovnio/wovnjava/HtmlConverterTest.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            "<!--wovn-marker-7-->" +
+            "<!--wovn-marker-8-->" +
+            "</body></html>";
+        Settings settings = makeSettings(new HashMap<String, String>() {{ put("supportedLangs", "en,fr,ja"); }});
+        HtmlConverter converter = new HtmlConverter(settings, original);
+        String html = converter.strip();
+        assertEquals(removedHtml, stripExtraSpaces(html));
+    }
+
+    private Settings makeSettings(HashMap<String, String> option) {
```
Comment (suplo) [inline src/test/java/com/github/wovnio/wovnjava/HtmlConverterTest.java]: duplicate code with ApiTest.java

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java (hunk 11 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: But the class and field is public.
So user program can access it.

File: src/test/java/com/github/wovnio/wovnjava/HtmlConverterTest.java (hunk 13 above)
Comment (hythof) [reply src/test/java/com/github/wovnio/wovnjava/HtmlConverterTest.java]: I moved the code to TestUtil.java

File: src/test/java/com/github/wovnio/wovnjava/ApiTest.java (hunk 12 above)
Comment (hythof) [reply src/test/java/com/github/wovnio/wovnjava/ApiTest.java]: I moved the code to TestUtil.java

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java (hunk 10 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: removed

File: src/main/java/com/github/wovnio/wovnjava/Settings.java (hunk 9 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/Settings.java]: fied

File: src/main/java/com/github/wovnio/wovnjava/Interceptor.java (hunk 8 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/Interceptor.java]: You are right.
I fixed it.

File: src/main/java/com/github/wovnio/wovnjava/Interceptor.java (hunk 7 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/Interceptor.java]: fixed

File: src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java (hunk 6 above)
Comment (hythof) [reply src/main/java/com/github/wovnio/wovnjava/HtmlConverter.java]: That's right.
I fixed it.
Thanks.

File: src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java (hunk 3 above)
Comment (mrasu) [reply src/main/java/com/github/wovnio/wovnjava/WovnServletFilter.java]: ok..

**Pull request merged**